File: java/src/main/java/generating/singleton/MainSingleton.java
package generating.singleton;

import java.util.Random;

class Singleton {
	private static Singleton instance;
	private Integer value = new Random().nextInt();
	private Singleton() {
	}
	public static Singleton getInstance() {
		if (instance == null) {
			synchronized (Singleton.class) {
				if (instance == null) {
					instance = new Singleton();
				}
			}
		}
		return instance;
	}
	public Integer getValue() {
		return this.value;
	}
}

public class MainSingleton {
	public static void main(String[] args) {
		Singleton singleton1 = Singleton.getInstance();
		System.out.println(singleton1.getValue());
		Singleton singleton2 = Singleton.getInstance();
		System.out.println(singleton2.getValue());
		System.out.println(singleton2 == singleton1);
	}
}

File: kotlin/src/main/kotlin/generating/singleton/MainSingleton.kt
package generating.singleton

import kotlin.random.Random

internal class Singleton private constructor() {
    val value: Int = Random.nextInt()

    companion object {
        var instance: Singleton? = null
            get() {
                if (field == null) {
                    synchronized(Singleton::class) {
                        if (field == null) {
                            field = Singleton()
                        }
                    }
                }
                return field
            }
            private set
    }
}

fun main() {
    val singleton1: Singleton = Singleton.instance!!
    println(singleton1.value)
    val singleton2: Singleton = Singleton.instance!!
    println(singleton2.value)
    println(singleton2 == singleton1)
}